src/Contract/UnbalancedTx.purs

```diff
@@ -35,24 +40,95 @@ import Ctl.Internal.ProcessConstraints.Error
       , TxOutRefNotFound
       , TxOutRefWrongType
       , WrongRefScriptHash
+      , ValidatorHashNotFound
       , CannotSatisfyAny
       , ExpectedPlutusScriptGotNativeScript
       , CannotMintZero
       )
   ) as X
 import Ctl.Internal.ProcessConstraints.UnbalancedTx (UnbalancedTx)
 import Ctl.Internal.ProcessConstraints.UnbalancedTx (UnbalancedTx(UnbalancedTx)) as X
+import Ctl.Internal.Transaction (explainModifyTxError)
+import Ctl.Internal.Types.Interval (explainPosixTimeToSlotError)
 import Ctl.Internal.Types.ScriptLookups
   ( ScriptLookups
   )
 import Ctl.Internal.Types.TxConstraints (TxConstraints)
-import Data.Either (Either)
+import Data.Either (Either(Left, Right))
+import Effect.Exception (error)
 
 -- | Create an `UnbalancedTx` given `ScriptLookups` and
 -- | `TxConstraints`. This should be called in conjuction with
 -- | `balanceTx` and  `signTransaction`.
+-- |
+-- | This is a 'non-throwing' variant; use this when failure is expected, and
+-- | can be handled gracefully.
 mkUnbalancedTx
   :: ScriptLookups
   -> TxConstraints
   -> Contract (Either MkUnbalancedTxError UnbalancedTx)
 mkUnbalancedTx = PC.mkUnbalancedTxImpl
+
+-- | As `mkUnbalancedTx`, but 'throwing'; use this when failure is _not_
+-- | expected, and graceful handling is either impossible, or wouldn't make
+-- | sense.
+mkUnbalancedTxE :: ScriptLookups -> TxConstraints -> Contract UnbalancedTx
+mkUnbalancedTxE lookups constraints =
+  mkUnbalancedTx lookups constraints >>= case _ of
+    Left err -> throwError $ error $ explainMkUnbalancedTxError err
+    Right res -> pure res
+
+-- | Helper to pretty-print `MkUnbalancedTxError`s.
+explainMkUnbalancedTxError :: MkUnbalancedTxError -> String
+explainMkUnbalancedTxError = case _ of
+  X.CannotConvertPaymentPubKeyHash ppkh ->
+    "Cannot convert payment pubkey hash: " <> show ppkh
+  X.CannotFindDatum -> "Cannot find datum."
+  X.CannotQueryDatum dh -> "Cannot query datum: " <> show dh
+  X.CannotConvertPOSIXTimeRange tr ttsErr ->
+    "Cannot convert POSIX time range: " <> show tr <> "\nReason: " <>
+      explainPosixTimeToSlotError ttsErr
+  X.CannotSolveTimeConstraints tr tr' ->
+    "Unsolvable time constraints: " <> show tr <> " and " <> show tr' <>
+      " do not overlap."
+  X.CannotGetMintingPolicyScriptIndex ->
+    "Cannot get minting policy script index. This should be impossible."
+  X.CannotGetValidatorHashFromAddress addr ->
+    "Cannot get a validator hash from address " <> show addr
+  X.CannotHashDatum datum ->
+    "Cannot hash datum: " <> show datum
+  X.CannotHashMintingPolicy mp ->
+    "Cannot hash minting policy: " <> show mp
+  X.CannotHashValidator validator ->
+    "Cannot hash validator: " <> show validator
+  X.CannotMakeValue cs tn amount ->
```

Comment:
The source is here:

```
      let value = mkSingletonValue' cs tn
      -- If i is negative we are burning tokens. The tokens burned must
      -- be provided as an input. So we add the value burnt to
      -- 'valueSpentBalancesInputs'. If i is positive then new tokens are
      -- created which must be added to 'valueSpentBalancesOutputs'.
      -- If i is zero we raise error, because of
      -- https://github.com/Plutonomicon/cardano-transaction-lib/issues/1156
      mintVal <-
        if i < zero then do
          v <- liftM (CannotMakeValue cs tn i) (value $ negate i)
          _valueSpentBalancesInputs <>= provideValue v
          pure $ map getNonAdaAsset $ value i
        else if i == zero then do
          throwError $ CannotMintZero cs tn
        else do
          v <- liftM (CannotMakeValue cs tn i) (value i)
          _valueSpentBalancesOutputs <>= provideValue v
          pure $ map getNonAdaAsset $ value i
```

This is used because `mkSingletonValue'` can fail if given the combination of the ADA `CurrencySymbol` and a non-empty `TokenName`. I will amend the error message to take this into account, but I'm pretty sure this should be impossible anyway.